### PR TITLE
Add bank conflicts information to perfRunner

### PR DESCRIPTION
**Pull request by stefankoncarevic**
I used rocprof to find out the bank conflict information for convolution and gemm. During the mlir benchmark in the output .csv file there is a column containing the results for all configurations. I also managed to find bank conflicts for rocblas benchmarking and write them into same output .csv file. The percentage of GPUTime LDS is stalled by bank
conflicts. Value range: 0% (optimal) to 100% (bad).

### CODE REVIEW

File: mlir/utils/performance/perfRunner.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 ROCPROF = '/opt/rocm/bin/rocprof'
 MIOPENDRIVER = '/opt/rocm/bin/MIOpenDriver'
 BENCHMARKING_RESULT_FILE_NAME = 'results.stats.csv'
+BANKCONFLICT = 'results.csv'
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: Can you use a more generic name, like `BENCHMARKING_METRICS_FILE_NAME`? In theory we may want to add more metrics to analyze (in addition to the bank conflicts)

File: mlir/utils/performance/perfRunner.py [hunk 2]
```diff
@@ -114,6 +116,7 @@ def create_paths(config_file_path, mlir_build_dir_path) -> Paths:
             libmlir_rocm_runtime_path =  llvm_lib_dir + '/libmlir_rocm_runtime.so',
             libconv_validation_wrappers_path = mlir_lib_dir + '/libconv-validation-wrappers.so',
             libmlir_runtime_utils_path = llvm_lib_dir + '/libmlir_runner_utils.so',
+            libmlir_c_runner_utils_path = llvm_lib_dir + '/libmlir_c_runner_utils.so',
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: Why you need the `c_runner`?

File: mlir/utils/performance/perfRunner.py [hunk 3]
```diff
@@ -135,6 +138,26 @@ def getNanoSeconds(fileName):
...
 
+# Bank conflict functions.The percentage of GPUTime LDS is stalled by bank
+# conflicts. Value range: 0% (optimal) to 100% (bad).
+def getBankConflict(fileName):
+    if not os.path.exists(fileName):
+        result = "NaN"
+        return result
+    with open(fileName, 'r') as csv_file:
+        reader = csv.DictReader(csv_file, delimiter = ',')
+
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py:153]: Nit: remove space

File: mlir/utils/performance/perfRunner.py [hunk 4]
```diff
@@ -135,6 +138,26 @@ def getNanoSeconds(fileName):
...
+        result = []
+        for row in reader:
+            if 'LDSBankConflict' in row:
+                result.append(float(row['LDSBankConflict']))
+            else:
+                return np.nan
+        csv_file.close()
+
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: Nit: remove space

File: mlir/utils/performance/perfRunner.py [hunk 5]
```diff
@@ -135,6 +138,26 @@ def getNanoSeconds(fileName):
...
+def getBankConflict(fileName):
+    if not os.path.exists(fileName):
+        result = "NaN"
+        return result
+    with open(fileName, 'r') as csv_file:
+        reader = csv.DictReader(csv_file, delimiter = ',')
+
+        result = []
+        for row in reader:
+            if 'LDSBankConflict' in row:
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: Can you check this once (outside the loop)? I mean, if the `LDSBankClonflict` is not in the header, then return `nan`

File: mlir/utils/performance/results.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+pmc: LDSBankConflict
```
Comment (giuseros) [inline mlir/utils/performance/results.txt]: Can you add a bit of comments explaining how this work, why this file is needed, etc..?

File: mlir/utils/performance/perfRunner.py [hunk 7]
```diff
@@ -1015,31 +1049,35 @@ def getFusionTestInfo(filename, paths: Paths):
...
+    # rocmlir-gen -fut test -arch gfx90a --clone-harness
+    rocmlirgenCommand = [paths.mlir_paths.rocmlir_gen_path, '-fut', futName, '-arch', getChip(), '--clone-harness', filename]
+    p0 = subprocess.Popen(rocmlirgenCommand, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+    if "-migraphx-to-tosa" in rocmlirCommand:
         rocmlirOptCommand = [paths.mlir_paths.rocmlir_opt_path, '-migraphx-to-tosa', filename]
-        rocmlirDriverCommand = [paths.mlir_paths.rocmlir_driver_path, '-host-pipeline', 'partition,highlevel', '-targets', getChip()]
+        rocmlirDriverCommand = [paths.mlir_paths.rocmlir_driver_path, '-host-pipeline', 'highlevel', '-targets', getChip()]
         # rocmlir-opt -migraphx-to-tosa ../mlir/test/fusion/resnet50-e2e/mixr-resnet-fusion-case-1.mlir
-        p1 = subprocess.Popen(rocmlirOptCommand, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+        p1 = subprocess.Popen(rocmlirOptCommand, stdin=p0.stdout, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: How is this related to this change?

Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: Ho this is related to your change?

File: mlir/utils/performance/perfRunner.py [hunk 8]
```diff
@@ -1115,7 +1153,7 @@ def benchmarkFusionKernels(test_dir, paths: Paths, arch, numCU, tuningDb: MaybeT
                continue;
 
         # Run fusion test
-        rocmlirGenArgs = ['-ph', '-fut='+futName, '--perf_config='+bestPerf, '-']
+        rocmlirGenArgs = ['-ph', '-fut='+futName+'_wrapper', '--perf_config='+bestPerf, '-']
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: How this is related to your change?

File: mlir/utils/performance/perfRunner.py [hunk 9]
```diff
@@ -1259,7 +1297,7 @@ def main(args=None):
     numCU = getNumCU(chip)
 
     root_dir = str(subprocess.check_output(['git', 'rev-parse', '--show-toplevel']).decode().strip())
-    default_conv_configs = root_dir + '/mlir/utils/jenkins/performance/conv-configs'
+    default_conv_configs = root_dir + '/mlir/utils/performance/conv-configs'
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: How this is related to your change?

**Review by giuseros — CHANGES_REQUESTED** [review]
I think there is some code that is not related to this change. Other than that, and few minor comments, the code structure looks ok

File: mlir/utils/performance/perfRunner.py (hunk 1 above)
Comment (stefankoncarevic) [reply mlir/utils/performance/perfRunner.py]: Yes, I will change that.

File: mlir/utils/performance/results.txt (hunk 6 above)
Comment (giuseros) [reply mlir/utils/performance/results.txt]: Also, I would call this file `rocmlir_metrics.txt` or something like this

File: mlir/utils/performance/perfRunner.py (hunk 2 above)
Comment (stefankoncarevic) [reply mlir/utils/performance/perfRunner.py]: I had conflicts because the perfrunner was changed and I was working on an older version, so then I just changed the parts to be like the last one.

File: mlir/utils/performance/results.txt (hunk 6 above)
Comment (stefankoncarevic) [reply mlir/utils/performance/results.txt]: Yes I will add the explaining for that.

File: mlir/utils/performance/perfRunner.py (hunk 2 above)
Comment (giuseros) [reply mlir/utils/performance/perfRunner.py]: I think you need to rebase your changes against `origin/develop`. If you are unsure, feel free to reach out to me. The main thing is that I should see only the changes relevant to your change. Other PRs need to be integrated by rebasing

**Comment by giuseros** [comment]
Good Job @stefankoncarevic ! Last question: does the `perfRegressionReport.py` still work? Does it report the difference in LDS bank conflicts as well?

**Comment by stefankoncarevic** [comment]
> Good Job @stefankoncarevic ! Last question: does the `perfRegressionReport.py` still work? Does it report the difference in LDS bank conflicts as well?

**Comment by stefankoncarevic** [comment]
Thank you so much @giuseros . I will check today, since @djramic changed the script on Friday, whether all the information is saved properly.

File: mlir/utils/performance/perfRunner.py [hunk 10]
```diff
@@ -27,6 +27,8 @@
 ROCPROF = '/opt/rocm/bin/rocprof'
 MIOPENDRIVER = '/opt/rocm/bin/MIOpenDriver'
 BENCHMARKING_RESULT_FILE_NAME = 'results.stats.csv'
+BENCHMARKING_METRICS_FILE_NAME = 'results.csv'
+ROCMLIR_INPUT_METRICS_FILE_NAME = 'results.txt'
```
Comment (giuseros) [inline mlir/utils/performance/perfRunner.py]: I would call this something like `rocmlir_metrics.txt`. This is the file containing the metrics, not the results

Comment (stefankoncarevic) [reply mlir/utils/performance/perfRunner.py]: Yes I have problem if change the name like (rocmlir_metrics.txt) in output file I don't have number of TFlops got NaN, because this line when call [ROCPROF, '-i', file_path, '--stats', paths.mlir_paths.cpu_runner_path] + mlir_cpu_runner_args will create rocmlir_metrics.stats.csv output file not results.stats.csv and that is a problem.

Comment (stefankoncarevic) [reply mlir/utils/performance/perfRunner.py]: With '-i' and '--stats' together will create a output file from input name. For this reason, I tried to rename everything to rocmlir_metrics, but then there is a problem with the tuning runner because it only calls '--stats' and by default results.csv text output files are created.

Comment (giuseros) [reply mlir/utils/performance/perfRunner.py]: Well you can specify the `-o` option with `rocprof`. I just tested this:
```
rocprof --stats  -i prof.txt -o mlir-prof.csv 
```
And I get a `mlir-prof.stats.csv` and a `mlir-prof.csv`. So I guess you need to do something like :
```
[ROCPROF, '-i', ROCMLIR_INPUT_METRICS_FILE_NAME , '--stats', paths.mlir_paths.cpu_runner_path, -o, BENCHMARKING_METRICS_FILE_NAME ] 
```
And I would stick with:
```
BENCHMARKING_RESULT_FILE_NAME = 'results.stats.csv'
BENCHMARKING_METRICS_FILE_NAME = 'results.csv'
ROCMLIR_INPUT_METRICS_FILE_NAME = 'rocmlir_metrics.txt'
```

Comment (stefankoncarevic) [reply mlir/utils/performance/perfRunner.py]: Thank you @giuseros, I will try to do that.

**Comment by giuseros** [comment]
Everything looks fine now, but I am wondering still how `perfRegression.py` is going to work. Can you try it and check what it prodcues?

**Review by giuseros — APPROVED** [review]
LGTM! Sorry for the pedantry :)

**Comment by stefankoncarevic** [comment]
Thank you very much for your help.
